### DISCUSSION TITLE
Iterate with JavaScript for Loops

### DISCUSSION THREAD

**Discussion by Skye404** [USU-ITLS-x265-23-summer/class-repo#121] [Q&A] [answered]
<img width="925" alt="Screenshot 2023-06-13 121326" src="https://github.com/USU-ITLS-x265-23-summer/class-repo/assets/133051222/f0f4e329-44be-4dd1-8b9f-80ca8ab2656a">
I've tried changing little things, but I'm still unable to get this correctly. I've watched the help video and looked online. Any ideas?

**Comment by metasean** [comment, marked as answer]
1. One of the first things I do when trying to find a bug in someone else's code is to run their existing code, but if you have a typo I may not actually notice the typo as I manually go back and forth between looking at your code and typing it in myself (it's sort of an unconscious auto-correct), and it's also easy for me to inadvertently add a typo, and spend time trying to find a bug that you don't actually have.

   So, when posting a bug, you should also post your code in a copyable format, e.g.
   ```js
   // Setup
   const myArray = [];
 
   // Only change code below this line
   for (var i=1; i < 5; i++) {
     myArray.push(i);
   }
   ```
   

4. There are multiple ways to debug code (you'll be learning some of them in an upcoming lesson), in the meantime, there are a couple of easy methods to figure out what your code is actually producing.  
   1. The most common way is to add a `console.log`; there are ways that this can really backfire, so there is a limit to it's utility.

      Adding a useful console log to your code (at least, your code as I manually copied it) might look like:
      ```js
      // Setup
      const myArray = [];

      // Only change code below this line
      for (var i=1; i < 5; i++) {
        myArray.push(i);
        console.log(i); // 👈
      }
      ``` 
      If you run this variation , the error should jump right out at you. 😉 

   2. A second way —that is great for single JS functions, or a set of small JS functions— is to use https://pythontutor.com/javascript.html  This will actually give you a view into what's happening at each step of the code.  If you click the "Visualize Execution" button on [a page with your code](https://pythontutor.com/javascript.html#code=//%20Setup%0Aconst%20myArray%20%3D%20%5B%5D%3B%0A%0A//%20Only%20change%20code%20below%20this%20line%0Afor%20%28var%20i%3D1%3B%20i%20%3C%205%3B%20i%2B%2B%29%20%7B%0A%20%20myArray.push%28i%29%3B%0A%7D&heapPrimitives=nevernest&mode=edit&origin=opt-frontend.js&py=js&rawInputLstJSON=%5B%5D) (again, as I manually copied it) you'll be taken to a page with your code on the left, and the visualization on the right.  You can use the "Next >" button (and related buttons) to "step through" your code.  
   
      In this case, I think the `console.log` is the faster way to identify the bug, but PythonTutor (running the JavaScript Tutor) can be really useful with other bugs, as well as just helping you understand what's happening in your the code.

5. You should banish `var` from your life.  It's a new dawn, and we only use `const` or `let` in these new days! 😉 

6. Adding what you learn from this process to the course-notes will probably be useful for future reference. 😉

**Reply by Skye404** [reply]
The console.log helped! I needed to add a 6 instead of a 5 there. I don't understand why though.
Thanks for the debugging resource too. I'll add these to the course-notes.

**Reply by coopsand** [reply]
@Skye404 You needed to add a 6 instead of a 5 because you were starting at 1 and not 0. You had it set that the variable could go up to **but not including** 5 (<), so it began at 1 and continued counting up to 4. (It looped 4 times.) When "i" became 5, it stopped the loop and didn't run the code inside. If you start it at 0, and didn't include 5 (using < and NOT <=) it would count 0, 1, 2, 3, 4 (5 times).

I don't know if that helped you out or not. Loops are a little tricky the first time you learn them. Please feel free to ask more questions and I can attempt to help you further.

**Reply by Skye404** [reply]
Super helpful! Thanks!

**Comment by coopsand** [comment]
Hey Skye!

I wanted to add on a couple more points to what @metasean said. I know he mentioned `console.log()` already, but I find it so incredibly useful, so I wanted to add a little more.

1. Using an online compiler, you can attempt to run your code. By using `console.log(*INSERT VAR HERE*)` you are able to actually print to the console what your code is doing. I took your code and put it inside [this](https://onecompiler.com/javascript/3zbkyda6k) compiler, and as opposed to pushing numbers into the array, **I used `console.log()` to see what the loop was doing.** That way, you'll know that the problem is/isn't something to do with the loop.
2. Here's another hint if you haven't quite figured it out yet: **Be careful with numerical values**.

I hope this helped.
